c/parallel/src/scan.cu

```diff
@@ -0,0 +1,534 @@
+  return init.type;
+}
+
+std::string get_input_iterator_name()
+{
+  std::string iterator_t;
+  check(nvrtcGetTypeName<input_iterator_state_t>(&iterator_t));
+  return iterator_t;
+}
+
+std::string get_output_iterator_name()
+{
+  std::string iterator_t;
+  check(nvrtcGetTypeName<output_iterator_t>(&iterator_t));
+  return iterator_t;
+}
+
+std::string
+get_init_kernel_name(cccl_iterator_t input_it, cccl_iterator_t /*output_it*/, cccl_op_t op, cccl_value_t init)
+{
+  const cccl_type_info accum_t  = scan::get_accumulator_type(op, input_it, init);
+  const std::string accum_cpp_t = cccl_type_enum_to_name(accum_t.type);
+  return std::format("cub::detail::scan::DeviceScanInitKernel<cub::ScanTileState<{0}>>", accum_cpp_t);
+}
+
+std::string get_scan_kernel_name(cccl_iterator_t input_it, cccl_iterator_t output_it, cccl_op_t op, cccl_value_t init)
+{
+  std::string chained_policy_t;
+  check(nvrtcGetTypeName<device_scan_policy>(&chained_policy_t));
+
+  const cccl_type_info accum_t  = scan::get_accumulator_type(op, input_it, init);
+  const std::string accum_cpp_t = cccl_type_enum_to_name(accum_t.type);
+  const std::string input_iterator_t =
+    (input_it.type == cccl_iterator_kind_t::pointer //
+       ? cccl_type_enum_to_name(input_it.value_type.type, true) //
+       : scan::get_input_iterator_name());
+  const std::string output_iterator_t =
+    output_it.type == cccl_iterator_kind_t::pointer //
+      ? cccl_type_enum_to_name(output_it.value_type.type, true) //
+      : scan::get_output_iterator_name();
+  const std::string init_t = cccl_type_enum_to_name(init.type.type);
+
+  std::string offset_t;
+  check(nvrtcGetTypeName<OffsetT>(&offset_t));
+
+  std::string scan_op_t;
+  check(nvrtcGetTypeName<op_wrapper>(&scan_op_t));
+
+  auto tile_state_t = std::format("cub::ScanTileState<{0}>", accum_cpp_t);
+  return std::format(
+    "cub::detail::scan::DeviceScanKernel<{0}, {1}, {2}, {3}, {4}, {5}, {6}, {7}, {8}, {9}>",
+    chained_policy_t,
+    input_iterator_t,
+    output_iterator_t,
+    tile_state_t,
+    scan_op_t,
+    init_t,
+    offset_t,
+    accum_cpp_t,
+    "false", // for now, always exclusive
+    init_t);
+}
+
+static constexpr auto ptx_u64_assignment_regex = R"(\.visible\s+\.global\s+\.align\s+\d+\s+\.u64\s+{}\s*=\s*(\d+);)";
+
+size_t find_size_t(char* ptx, std::string_view name)
+{
+  std::regex regex(std::format(ptx_u64_assignment_regex, name));
+  std::cmatch match;
+  if (std::regex_search(ptx, match, regex))
+  {
+    auto result = std::stoi(match[1].str());
+    return result;
+  }
+  else
+  {
+    throw std::runtime_error(std::format("Could not find {} in PTX code", name));
+  }
+}
+
+size_t find_size_t(char* ptx, std::string_view name, size_t default_value)
+{
+  std::regex regex(std::format(ptx_u64_assignment_regex, name));
+  std::cmatch match;
+  if (std::regex_search(ptx, match, regex))
+  {
+    return std::stoi(match[1].str());
+  }
+  else
+  {
+    return default_value;
+  }
+}
```

Comment:
Could you use `std::optional<size_t>` here to avoid the code duplication?

If `std::nullopt` throw, otherwise `return default_value;`

Reply:
Great suggestion. I decided to do away with  the overloads and instead have this function return `nullopt` when the search comes up empty. The caller is free to do what they like when a `nullopt` is returned.

This change is implemented in https://github.com/NVIDIA/cccl/pull/3462/commits/c3f719f4c5c7d934fc31b33e1f1c034ddf5e631f.